### PR TITLE
feat(beahviors): Allow encoders to use any behavior

### PR DESCRIPTION
**Pull request by nickconway**
This PR adds the sensor-rotate behavior, which will allow for encoders to use any behavior in addition to basic keypresses.

In the documentation I have assumed this will deprecate the existing &inc_dec_kp behavior.

This has not been tested extensively but is confirmed to be working with RGB behaviors.

### CODE REVIEW

**Comment by eXsoR65** [comment]
Working with rgb behavior:

https://user-images.githubusercontent.com/19521349/165663211-e25bb3d1-510c-4e8e-abd8-4b43eee00fb1.mov

File: app/dts/bindings/behaviors/zmk,behavior-sensor-rotate.yaml
Comment (kurtis-lew) [inline app/dts/bindings/behaviors/zmk,behavior-sensor-rotate.yaml:1]: Copyright headers should be updated to the current year (2022)

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (kurtis-lew) [inline app/src/behaviors/behavior_sensor_rotate.c:2]: Same thing for updating the year of the copyright header as the devicetree binding (`.yaml`)

File: docs/docs/behaviors/sensor-rotate.md
Comment (kurtis-lew) [inline docs/docs/behaviors/sensor-rotate.md:21]: `zmk,behavior-mod-morph` should be replaced with `zmk,behavior-sensor-rotate`

File: docs/docs/features/encoders.md
Comment (kurtis-lew) [inline docs/docs/features/encoders.md:36]: From the `behaviors/sensor-rotate.md` example, users can find the RGB control behavior that `&rgb_encoder` refers to. However, it's not immediately clear that the `&arrow_encoder` behavior is bound to the left/right arrow keys on rotation.
On that note, if the updated sensor-rotate behavior does deprecate the sensor-rotate-kp behavior, would it be more appropriate to combine the main page, `encoders.md`, with the new behavior page you've created in this PR? If so, I'd think a code block with the `&rgb_encoder` and `&arrow_encoder` sensor-rotate behaviors would fit really well here.

**Review by kurtis-lew — COMMENTED** [review]
Looking good so far! I'm currently unable to get this feature working on my local hardware, but in the meantime I have a few comments.

**Comment by kurtis-lew** [comment]
image

Another thing that I'm picking up when I build firmware locally using this PR is the error in the image above. In my case, when I rotate my encoder clockwise or counterclockwise, I'm unable to get the correct behavior binding to register. Perhaps the incompatible pointer types may be the reason for this issue. I'll continue to debug things from my end and see if I can confirm this.

**Comment by eXsoR65** [comment]
> image
> 
> Another thing that I'm picking up when I build firmware locally using this PR is the error in the image above. In my case, when I rotate my encoder clockwise or counterclockwise, I'm unable to get the correct behavior binding to register. Perhaps the incompatible pointer types may be the reason for this issue. I'll continue to debug things from my end and see if I can confirm this.

I too get the same warring during the build process though Its still working. 
https://github.com/eXsoR65/bdn9_zmk_config/runs/6198384676?check_suite_focus=true#step:9:65


Here my config: https://github.com/eXsoR65/bdn9_zmk_config

File: docs/docs/features/encoders.md
Comment (kurtis-lew) [reply docs/docs/features/encoders.md:36]: Another thing I just thought of: if the final decision is to keep the separate `docs/behaviors/sensor-rotate.md`, the behaviors section of `docs/sidebars.js` should also be updated to include the new page, enhancing accessibility.

File: app/dts/bindings/behaviors/zmk,behavior-sensor-rotate.yaml
Comment (nickconway) [reply app/dts/bindings/behaviors/zmk,behavior-sensor-rotate.yaml]: Whoops! Fixed these.

File: docs/docs/features/encoders.md
Comment (nickconway) [reply docs/docs/features/encoders.md:56]: I've added the other encoder into the example code for now.

**Review by kurtis-lew — APPROVED** [review]

File: docs/docs/features/encoders.md
Comment (kurtis-lew) [inline docs/docs/features/encoders.md:56]: Similar note as that for `sensor-rotate.md` for capitalization, but also enhancing clarity on the behaviour of the `arrow_encoder` binding.

```suggestion
Here, the left encoder is configured to control RGB brightness while the right encoder sends either the left or right arrow keys.
```

File: docs/docs/behaviors/sensor-rotate.md
Comment (kurtis-lew) [inline docs/docs/behaviors/sensor-rotate.md:15]: Another minor detail in passing: suggesting capitalization of "RGB" for consistency across the docs.
```suggestion
An example implementation of an encoder that changes RGB brightness is shown below:
```

**Review by kurtis-lew — CHANGES_REQUESTED** [review]
Did another pass through this PR to make some more minor suggestions. Thanks for the updates based on the first review!

File: docs/docs/behaviors/sensor-rotate.md
Comment (drzony) [inline docs/docs/behaviors/sensor-rotate.md:8]: Maybe change "sends a different keypress" into "triggers a configured behavior"

File: docs/docs/behaviors/sensor-rotate.md
Comment (drzony) [inline docs/docs/behaviors/sensor-rotate.md:10]: Instead of "keycode is sent" maybe "behavior is triggered"

File: docs/docs/behaviors/sensor-rotate.md
Comment (nickconway) [reply docs/docs/behaviors/sensor-rotate.md]: I think this wording makes more sense and I've added it to the pr, thanks for the suggestion!

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (joelspadin) [inline app/src/behaviors/behavior_sensor_rotate.c:20]: There are some recent improvements to Zephyr (documented in https://docs.zephyrproject.org/2.6.0/guides/kconfig/preprocessor-functions.html) that provide a more efficient way to do this.

First add to app/Kconfig something like
```
DT_COMPAT_ZMK_BEHAVIOR_SENSOR_ROTATE := zmk,behavior-sensor-rotate

config ZMK_BEHAVIOR_SENSOR_ROTATE
    bool
    default $(dt_compat_enabled,$(DT_COMPAT_ZMK_BEHAVIOR_SENSOR_ROTATE))
```

and then in CMakeLists.txt you can use
```cmake
target_sources_ifdef(CONFIG_ZMK_BEHAVIOR_SENSOR_ROTATE app PRIVATE src/behaviors/behavior_sensor_rotate.c)
```
so that this whole file just gets excluded from the build if the behavior isn't used anywhere. Then you don't need to conditionally compile the contents of it.

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (joelspadin) [inline app/src/behaviors/behavior_sensor_rotate.c:65]: You should use `zmk_behavior_queue_add()` so you can queue the press and release without blocking the main thread.

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (joelspadin) [inline app/src/behaviors/behavior_sensor_rotate.c:86]: Minor: I'm guessing you copied this from the key press behavior? The naming of this doesn't make a lot of sense here.

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (nickconway) [reply app/src/behaviors/behavior_sensor_rotate.c:65]: Should the `k_msleep(5)` be kept in between the press and release with this method?

Comment (joelspadin) [reply app/src/behaviors/behavior_sensor_rotate.c:65]: No. The point is to allow scheduling the release to happen later without forcing the main thread to do nothing for 5 milliseconds. You can do

```c
zmk_behavior_queue_add(position, &binding, true, 0);
zmk_behavior_queue_add(position, &binding, false, 5);
```

I am not sure how to set the key position for a sensor though. The existing code here just uses 0 (by way of not setting the position when initializing `event`).

(We should probably also make a similar change to behavior_sensor_rotate_key_press.c in a separate PR.)

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (nickconway) [reply app/src/behaviors/behavior_sensor_rotate.c:81]: I think I copied from mod-morph but it seems like most of the behaviors use KP e.g. tap dance.

Comment (kurtis-lew) [reply app/src/behaviors/behavior_sensor_rotate.c:81]: This is something that will be addressed in an upcoming PR. Functionally, it doesn't matter what the `XX_INST(n)` is labelled as, but consistency with the behavior name is nice to have.

Comment (nickconway) [reply app/src/behaviors/behavior_sensor_rotate.c]: Changed the name to match the behavior.

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (petejohanson) [reply app/src/behaviors/behavior_sensor_rotate.c]: Fully agreed on this, that API/infrastructure didn't exist when the original sensor behavior was written, which I imagine this was copied from.

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate.c:64]:
```suggestion
```

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate.c:60]: The 5 here is arbitrary, and not great. Maybe we should make this an optional value in the DT, that defaults to 5 to keep the existing functionality unchanged?

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate.c:48]: We bother storing this to local data? Seems like this can just be a local variable we set, then pass to the queue. Then we can skip the data altogether for this driver, just use the const config.

File: docs/docs/features/encoders.md
Comment (petejohanson) [inline docs/docs/features/encoders.md:45]: Looking at this...  I'm not sure I love how much this complicates the simple case. It might be really nice to have an equivalent generic behavior (this one, but w/ tweaks?) that takes a binding cells value of 2, and bindings *without* params, and forwards on the params bound to it.

It would allow us to redefine `inc_dec_kp` like:

```
/ {
	behaviors {
		/* DEPRECATED: `inc_dec_cp` will be removed in the future */
		/omit-if-no-ref/ inc_dec_cp: inc_dec_kp: behavior_sensor_rotate_key_press {
			compatible = "zmk,behavior-sensor-rotate-var";
			label = "ENC_KEY_PRESS";
			#sensor-binding-cells = <2>;
                        bindings = <&kp>, <&kp>;
		};
	};
};
```

Thoughts?

**Review by petejohanson — CHANGES_REQUESTED** [review]
Thanks for working on this! A few minor comments, and one higher level change we might want to roll into this.

File: docs/docs/features/encoders.md
Comment (nickconway) [reply docs/docs/features/encoders.md]: Added this to the pr

File: app/src/behaviors/behavior_sensor_rotate_var.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate_var.c:59]: I'm not convinced this is actually doing what you want. It seems like this really should be:

```suggestion
    zmk_behavior_queue_add(0, triggered_binding, true, cfg->tap_ms);
    zmk_behavior_queue_add(0, triggered_binding, false, 0);
```

File: app/src/behaviors/behavior_sensor_rotate_var.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate_var.c:73]:
```suggestion
        .tap_ms = DT_INST_PROP(n, tap_ms),                                               \
```

File: app/dts/bindings/behaviors/zmk,behavior-sensor-rotate.yaml
Comment (petejohanson) [inline app/dts/bindings/behaviors/zmk,behavior-sensor-rotate.yaml:19]:
```suggestion
  tap-ms:
```

File: app/dts/bindings/behaviors/zmk,behavior-sensor-rotate-var.yaml
Comment (petejohanson) [inline app/dts/bindings/behaviors/zmk,behavior-sensor-rotate-var.yaml:19]:
```suggestion
  tap-ms:
```

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate.c:57]:
```suggestion
    zmk_behavior_queue_add(0, *triggered_binding, true, cfg->tap_ms);
    zmk_behavior_queue_add(0, *triggered_binding, false, 0);
```

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate.c:80]:
```suggestion
        .tap_ms = DT_INST_PROP_OR(n, tap_ms, 5),                                         \
```

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate.c:24]:
```suggestion
    int tap_ms;
```

File: app/src/behaviors/behavior_sensor_rotate_var.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate_var.c:24]:
```suggestion
    int tap_ms;
```

**Review by petejohanson — CHANGES_REQUESTED** [review]
Comment on the behavior queue usage, naming tweaks.

I don't love the code duplication between the parameterized version and not, but the amount duplicated is so minimal, I'd say it's fine.

**Comment by nickconway** [comment]
> I don't love the code duplication between the parameterized version and not, but the amount duplicated is so minimal, I'd say it's fine.

We could the duplicated code to another file that gets included in each behavior?

File: app/src/behaviors/behavior_sensor_rotate_var.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate_var.c:59]: Instead of simulating this at the zeroth key position, perhaps better to simulate it at "max key position + sensor position" so there are actually unique? May be necessary for some code that tracks state tied to key position (or might actually break them.... hmmm...)

**Review by petejohanson — COMMENTED** [review]
One more thought...

**Comment by ReFil** [comment]
Excited to see this as it will make the defeault layer shifting stuff i'vebeen working on much neater. Iimplemented a rudimentary sensor behaviour but universall sensor beaviours are even better

File: app/src/keymap.c
Comment (petejohanson) [inline app/src/keymap.c:272]: Thoughts on using `sensor_number` for the `position` value?

```suggestion
            struct zmk_behavior_binding_event event = {.position = sensor_number, .timestamp = timestamp};
```

File: app/src/behaviors/behavior_sensor_rotate_var.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate_var.c:39]:
```suggestion
    if (err < 0) {
```

We've been bit before by some Zephyr APIs that change to return positive numbers for some success conditions, so check for negative specifically is safer.

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (petejohanson) [inline app/src/behaviors/behavior_sensor_rotate.c:39]:
```suggestion
    if (err < 0) {
```

**Review by petejohanson — COMMENTED** [review]
Couple minor things. Want to give this a test before merging though.

Thanks for the work on this!

File: app/src/keymap.c
Comment (nickconway) [reply app/src/keymap.c:272]: I was just thinking to add something like this earlier, we'd need something to differentiate between sensors

**Comment by petejohanson** [comment]
@nickconway can you please rebase to pull in the docs build fixes?

**Comment by nickconway** [comment]
@petejohanson done.

File: docs/docs/behaviors/sensor-rotate.md
Comment (petejohanson) [inline docs/docs/behaviors/sensor-rotate.md:16]: We need an example for the -var and non-var variants here.

File: app/src/behaviors/behavior_sensor_rotate_var.c
Comment (petejohanson) [reply app/src/behaviors/behavior_sensor_rotate_var.c:61]: This got fixed in the non-var version, but not this one.

**Review by petejohanson — CHANGES_REQUESTED** [review]
A few minor fixes needed.

Also, we probably need to remove `app/dts/bindings/behaviors/zmk,behavior-sensor-rotate-key-press.yaml` in this PR as well as the driver code itself.

**Comment by thilojaeggi** [comment]
A bit off topic but I added this to my keymap file like so:
`
	sensors {
		sensors = <&left_encoder &right_encoder>;
	};
	
	behaviors {
        rgb_encoder: rgb_encoder {
            compatible = "zmk,behavior-sensor-rotate";
            label = "RGB_ENCODER";
            #sensor-binding-cells = <0>;
            bindings = <&rgb_ug RGB_BRD>, <&rgb_ug RGB_BRI>;
        };
    };
`

but it's not working.

What am I missing here?

File: app/src/behaviors/behavior_sensor_rotate.c
Comment (joelspadin) [inline app/src/behaviors/behavior_sensor_rotate.c:58]: I think this will overlap with combo positions?

**Comment by petejohanson** [comment]
Merged in the updated PR #1758

**Pull request closed without merging**
